### PR TITLE
Invalid crypto engine on OS X Mountain Lion (Safari, Chrome, and FF)

### PR DESCRIPTION
**Pull request by rca**
I built the plugin per the instructions, specifically, using:

``` bash
$ git clone git://github.com/kylehuff/webpg-npapi.git
$ cd webpg-npapi/
$ git submodule update --init --recursive
$ firebreath/prepmac.sh . build -DCMAKE_BUILD_TYPE=MinSizeRel
$ cd build/
$ xcodebuild -configuration MinSizeRel -target webpgPlugin
```

Then I copied the plugin `build/projects/webpgPlugin/MinSizeRel/npwebpgPlugin-v0.6.1.plugin` into `~/Library/Internet Plugins/`.

In all three NPAPI browsers, when I load the page `plugin_tests/various_tests.html` and call `plugin.getPublicKeyList()` the following object is returned:

``` javascript
{
error:true,
error_string:"Invalid crypto engine",
file:"/Users/berto/Projects/webpg-npapi/firebreath/../webpgPlugin/webpgPluginAPI.cpp", gpg_error_code:150,
line:664,
method:"getKeyList"
}
```

I found Issue #19, which suggests installing `gnupg2`.  Initially I didn't have it installed so I installed it from source and it landed in `/usr/local/bin/gpg2`.  When it didn't fix the problem I rebuilt the plugin.  That didn't fix the problem either.  I also symlinked gpgconf into `/usr/bin` just in case, but that didn't fix it either.

In Chrome when I call the function `plugin.gpgGenKey('RSA', 4096, 'RSA', 4096, 'Test', 'Test', 'test@foo.com', 0, 'test')` I get a warning banner that says, "WebPG has crashed.".  When calling that in Safari subsequent calls to the plugin state, "TypeError: 'undefined' is not a function (evaluating 'plugin.getPublicKeyList()')".

Any ideas?  Thanks!

### CODE REVIEW

**Comment by kylehuff** [comment]
At this point, I have no ideas as to the actual problem, however, if I can suggest that you install the browser extension webpg-chrome, and see if you have the same result. If it works, then I suspect something is awry with the build environment, or there is a missing dependency...

You can download the browser extension webpg-chrome from the google chrome webstore: https://chrome.google.com/webstore/detail/webpg/hhaopbphlojhnmbomffjcbnllcenbnih

**Comment by rca** [comment]
Wasn't a build issue.  I've not used `gpg` on this Mac so didn't have any of the tools installed.  In order to build the plugin I followed the the README's instructions to install the needed libraries, but it didn't mention I needed the tools installed in order for it to work (I assumed the libraries provided the functionality).

I cleaned up the README a bit and added a Prerequisites section in 6d573fe.  Hopefully this helps other people get started without fumbling like I did.

**Comment by rca** [comment]
By the way, note that I did build gnupg from source on my machine.  When I tried using gpg from the command line I encountered an error about not having a `pinentry` program.

I could not find how to simply build pinentry for OS X so I turned to the [GPGTools](https://www.gpgtools.org/) project.  After installing that, the plugin's test page started working.

**Comment by kylehuff** [comment]
Great, I'm glad you got it working. I thought you mentioned you had already installed GnuPG, but on second look, that was only GnuPG v2, which is only used for it's gpgconf utility.

As for the pinentry problem on OSX, the only way I have found to make it work is to use GPGTools, at least in Snow Leopard.

**Comment by rca** [comment]
Yep, GPGTools did the trick.  This plugin would be much better if it provided everything it needed out of the box.  Any way to package gpg as well?

**Comment by kylehuff** [comment]
That might make some users leery, as the packaged version of GnuPG could have a security flaw either intentional or accidental. I think at least for the time being, making it a dependency will have to suffice.

I will however add the presence of GnuPG (v1) part of my initial checks, and provide some better error reporting. The webpg-chrome extension currently does this for GnuPG v2.

Thanks for all of your contributions to solving this issue, it is greatly appreciated!

**Comment by rca** [comment]
To use webpg-npapi you expect the user to install a _browser_ extension that makes the entire system potentially accessible from any website out on the Internet and you're saying they're going to be leery about a packaged version of GnuPG?!

**Comment by kylehuff** [comment]
The likelihood of the sand-boxed and protected extension exposing the system (or even it's own methods) is unlikely, however, it is entirely likely that a new security vulnerability for GnuPG could arise, and protection from that would be entirely dependent on the ability of the extension maintainer to update their source and publish a new version, then the user would have actually update the extension.

It also introduces an additional security vulnerability whereby an attacker only needs to replace the GnuPG binary within the users local extension directory with a malicious version, which would not normally require any administrative rights.

Aside from the potential security risks with distributing binary executables, it would also hamper development and maintenance to need to package compatible versions for each of the supported operating systems, and to actually make them available to the system. No doubt the user would expect to be able to use GnuPG outside of the extension, which would be a nightmare.

webpg is an adjunct to an existing application, the expectation is that the user have that application -- I don't believe that to be unreasonable. If we started down that path, users would begin to expect that installing an extension (an adjunct to the browser) would also install the browser... =c )

In summary - I agree that it would be handy to have the GnuPG binary distributed with the extension, I just don't see it as a viable option at this point in time due to the complexity and potential vulnerabilities it introduces.
